transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,579 @@
+        })
+    }
+}
+
+/// Object containing all the QUIC resources shared between all connections.
+// TODO: expand docs
+// TODO: Debug trait
+// TODO: remove useless fields
+pub struct Endpoint {
+    /// Channel to the background of the endpoint.
+    /// See [`Endpoint::new_connections`] (just below) for a commentary about the mutex.
+    to_endpoint: Mutex<mpsc::Sender<ToEndpoint>>,
+
+    /// Channel where new connections are being sent.
+    /// This is protected by a futures-friendly `Mutex`, meaning that receiving a connection is
+    /// done in two steps: locking this mutex, and grabbing the next element on the `Receiver`.
+    /// The only consequence of this `Mutex` is that multiple simultaneous calls to
+    /// [`Endpoint::next_incoming`] are serialized.
+    new_connections: Mutex<mpsc::Receiver<Connection>>,
+
+    /// Copy of [`Endpoint::to_endpoint`], except not behind a `Mutex`. Used if we want to be
+    /// guaranteed a slot in the messages buffer.
+    to_endpoint2: mpsc::Sender<ToEndpoint>,
+
+    /// Socketaddr of the local UDP socket passed in the configuration at initialization after it
+    /// has potentially been modified to handle port number `0`.
+    pub(crate) local_addr: SocketAddr,
+}
+
+impl Endpoint {
+    /// Builds a new `Endpoint`.
+    pub fn new(config: Config) -> Result<Arc<Endpoint>, io::Error> {
+        let local_socket_addr = match crate::transport::multiaddr_to_socketaddr(&config.multiaddr) {
+            Some(a) => a,
+            None => panic!(), // TODO: Err(TransportError::MultiaddrNotSupported(multiaddr)),
+        };
+
+        // NOT blocking, as per man:bind(2), as we pass an IP address.
+        let socket = std::net::UdpSocket::bind(&local_socket_addr)?;
+        // TODO:
+        /*let port_is_zero = local_socket_addr.port() == 0;
+        let local_socket_addr = socket.local_addr()?;
+        if port_is_zero {
+            assert_ne!(local_socket_addr.port(), 0);
+            assert_eq!(multiaddr.pop(), Some(Protocol::Quic));
+            assert_eq!(multiaddr.pop(), Some(Protocol::Udp(0)));
+            multiaddr.push(Protocol::Udp(local_socket_addr.port()));
+            multiaddr.push(Protocol::Quic);
+        }*/
+
+        let (to_endpoint_tx, to_endpoint_rx) = mpsc::channel(32);
+        let to_endpoint2 = to_endpoint_tx.clone();
+        let (new_connections_tx, new_connections_rx) = mpsc::channel(1);
+
+        let endpoint = Arc::new(Endpoint {
+            to_endpoint: Mutex::new(to_endpoint_tx),
+            to_endpoint2,
+            new_connections: Mutex::new(new_connections_rx),
+            local_addr: socket.local_addr()?,
+        });
+
+        // TODO: just for testing, do proper task spawning
+        async_global_executor::spawn(background_task(
+            config,
+            Arc::downgrade(&endpoint),
+            async_io::Async::<UdpSocket>::new(socket)?,
+            new_connections_tx,
+            to_endpoint_rx.fuse(),
+        ))
+        .detach();
```

Comment:
Even with #2652 we have to spawn a separate task so that we don't burden the main loop. But I think we can add `Config::executor` similar to [`SwarmBuilder::executor`](https://docs.rs/libp2p/latest/libp2p/swarm/struct.SwarmBuilder.html#method.executor).

Reply:
Solved with c5fc181.